pennylane/gradients/parameter_shift_cv.py

```diff
@@ -0,0 +1,678 @@
+        raise ValueError(
+            "Computing the gradient of circuits that return the state is not supported."
+        )
+
+    _gradient_analysis(tape)
+
+    gradient_tapes = []
+    shapes = []
+    fns = []
+
+    def _update(data):
+        """Utility function to update the list of gradient tapes,
+        the corresponding number of gradient tapes, and the processing functions"""
+        gradient_tapes.extend(data[0])
+        shapes.append(len(data[0]))
+        fns.append(data[1])
+
+    # TODO: replace the JacobianTape._grad_method_validation
+    # functionality before deprecation.
+    diff_methods = tape._grad_method_validation("analytic" if fallback_fn is None else "best")
+    all_params_grad_method_zero = all(g == "0" for g in diff_methods)
+
+    if not tape.trainable_params or all_params_grad_method_zero:
+        return gradient_tapes, lambda _: np.zeros([tape.output_dim, len(tape.trainable_params)])
+
+    # TODO: replace the JacobianTape._choose_params_with_methods
+    # functionality before deprecation.
+    method_map = dict(tape._choose_params_with_methods(diff_methods, argnum))
+    var_present = any(m.return_type is qml.operation.Variance for m in tape.measurements)
+
+    unsupported_params = [idx for idx, g in method_map.items() if g == "F"]
+    first_order_params = [idx for idx, g in method_map.items() if g == "A"]
+    second_order_params = [idx for idx, g in method_map.items() if g == "A2"]
+
+    if force_order2:
+        # all analytic parameters should be computed using the second-order method
+        second_order_params += first_order_params
+        first_order_params = []
+
+    if "PolyXP" not in dev.observables and (second_order_params or var_present):
+        warnings.warn(
+            f"The device {dev.short_name} does not support "
+            "the PolyXP observable. The analytic parameter-shift cannot be used for "
+            "second-order observables; falling back to finite-differences.",
+            UserWarning,
+        )
+
+        if var_present:
+            unsupported_params += first_order_params
+            first_order_params = []
+
+        unsupported_params += second_order_params
+        second_order_params = []
+
+    # If there are unsupported operations, call the callback gradient function
+    if unsupported_params:
+        _update(fallback_fn(tape, argnum=unsupported_params))
+
+    # collect all the analytic parameters
+    argnum = first_order_params + second_order_params
+
+    if not argnum:
+        # No analytic parameters. Return the existing fallback tapes/fn
+        return gradient_tapes, fns[-1]
+
+    gradient_recipes = gradient_recipes or [None] * len(argnum)
+
+    if var_present:
+        _update(var_param_shift(tape, dev.wires, argnum, shift, gradient_recipes, f0))
+
+    else:
+        if first_order_params:
+            _update(expval_param_shift(tape, first_order_params, shift, gradient_recipes, f0))

```

Comment:
How come `dev.wires` is not required for `expval_param_shift`?

Reply:
It's a unique requirement for the CV rule, since the transformation of the observable requires the device wires

Reply:
It is very annoying, I must say :( This is inherited from `CVParamShiftTape`